### ISSUE TITLE
C10_CUDA_KERNEL_LAUNCH_CHECK() becomes a no-op

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#91758]
### 🐛 Describe the bug

`C10_CUDA_KERNEL_LAUNCH_CHECK` calls `cudaGetLastError`:

https://github.com/pytorch/pytorch/blob/18b37bbff91d5746e5a1ec63bf82ce24a717434f/c10/cuda/CUDAException.h#L73

however, the result is discarded!

https://github.com/pytorch/pytorch/blob/18b37bbff91d5746e5a1ec63bf82ce24a717434f/c10/cuda/CUDAException.h#L27-L32

The code claims that the error will be obtained in `c10_cuda_check_implementation`, but that's not always true: many errors are non-sticky, meaning that after `cudaGetLastError` the error state is reset. In that case, this macro becomes a no-op.

In particular, launching a kernel with too many resources (too many registers & grid/block size) produces an error that's silently ignored by this macro -- and the kernel is silently not run :cry:  . Would be good to create a unittest about catching such errors.

It seems this bug was introduced in https://github.com/pytorch/pytorch/pull/85256 cc @ezyang @gchanan @zou3519 @ngimel @r-barnes .

### Versions

n/a

**Comment by ezyang** [comment]
@r-barnes when do you think you'll be able to follow up on this?

**Comment by malfet** [comment]
assigning to fix it unless @r-barnes will do it by EOW

**Comment by malfet** [comment]
Per @ngimelt this affected 1.13.0 and we should create a smoke check to make sure this would not happen in the future.
AI: @atalman to add a test

**Comment by d4l3k** [comment]
For SEO purposes since this wasn't searchable:

This breaks the expected of CUDA_LAUNCH_BLOCKING=1 for v1.13.1.

Patch for v1.13.1:

```patch
From e0576465a713bcc08625dacdf400368315950327 Mon Sep 17 00:00:00 2001
From: Tristan Rice <tristan.rice@getcruise.com>
Date: Thu, 27 Apr 2023 14:53:09 -0700
Subject: [PATCH] CUDAException: backport CUDA_LAUNCH_BLOCKING fix for v1.13.1

We can't directly cherry-pick the commits given other refactors but this
is based on two fixes in PT v2.0.

Source commits:
https://github.com/pytorch/pytorch/commit/eadbf762fc36f23ef78b084973dc39a13605db46
https://github.com/pytorch/pytorch/commit/f40183d374bc4f0cd7cf05e48915a5912214c564
---
 c10/cuda/CUDAException.cpp | 12 +++++++-----
 c10/cuda/CUDAException.h   |  2 ++
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/c10/cuda/CUDAException.cpp b/c10/cuda/CUDAException.cpp
index d35d72c9ba..29e1e0955e 100644
--- a/c10/cuda/CUDAException.cpp
+++ b/c10/cuda/CUDAException.cpp
@@ -9,22 +9,24 @@ namespace c10 {
 namespace cuda {
 
 void c10_cuda_check_implementation(
+    const int32_t err,
     const std::string& filename,
     const std::string& function_name,
     const int line_number,
     const bool include_device_assertions) {
-  // We retrieve the error here in order to keep CUDA data types out of
-  // CUDAException.h thereby simplifying including it in other files
-  const cudaError_t err = cudaGetLastError();
+  const auto cuda_error = static_cast<cudaError_t>(err);
 
-  if (C10_LIKELY(err == cudaSuccess)) {
+  if (C10_LIKELY(cuda_error == cudaSuccess)) {
     return;
   }
 
+  auto error_unused C10_UNUSED = cudaGetLastError();
+  (void)error_unused;
+
   std::string check_message;
 #ifndef STRIP_ERROR_MESSAGES
   check_message.append("CUDA error: ");
-  check_message.append(cudaGetErrorString(err));
+  check_message.append(cudaGetErrorString(cuda_error));
   check_message.append(c10::cuda::get_cuda_check_suffix());
 #endif
 
diff --git a/c10/cuda/CUDAException.h b/c10/cuda/CUDAException.h
index cfc7424503..2fcff615b9 100644
--- a/c10/cuda/CUDAException.h
+++ b/c10/cuda/CUDAException.h
@@ -27,6 +27,7 @@ class C10_CUDA_API CUDAError : public c10::Error {
     const cudaError_t __err = EXPR;                                        \
     if (C10_UNLIKELY(__err != cudaSuccess)) {                              \
       c10::cuda::c10_cuda_check_implementation(                            \
+          static_cast<int32_t>(__err),                                     \
           __FILE__,                                                        \
           __func__, /* Line number's data type is not well-defined between \
                        compilers, so we perform an explicit cast */        \
@@ -76,6 +77,7 @@ namespace cuda {
 /// In the event of a CUDA failure, formats a nice error message about that
 /// failure and also checks for device-side assertion failures
 C10_CUDA_API void c10_cuda_check_implementation(
+    const int32_t err,
     const std::string& filename,
     const std::string& function_name,
     const int line_number,
-- 
2.40.0
```
